ouroboros-network/src/Ouroboros/Network/Protocol/VersionNegotiation/Type.hs

```diff
@@ -0,0 +1,424 @@
+       ver
+    -> VersionData ver verData r
+    -> ProtocolVersion ver r
+
+-- |
+-- The @'Eq'@ instance ignores the continuation, it checks if version numbers
+-- are the same and verData have the same type and are equal.
+--
+instance Eq verNumber => Eq (ProtocolVersion verNumber r) where
+  ProtocolVersion ver (VersionData verData _) == ProtocolVersion ver' (VersionData verData' _) =
+    ver == ver' && Just verData == cast verData'
+
+-- |
+-- Data associated with a version.  Its type is existenially quantified in
+-- @'ProtocolVersion'@.  We put @Show@, @Serialise@ and @Typeable@ constraints
+-- needed for running the @'versionNegotiationClient'@ and
+-- @'versionNegotiationServer'@.
+--
+-- It is assumed that @ver@ implies the type of the existencially quantified
+-- @verData@ parameter, otherwise it is a protocol error
+-- @'VersionNegotiationTypeCastError'@.  @Typeable@ constraint is used for safe
+-- casting.
+--
+data VersionData ver verData r where
+  VersionData
+    :: ( Show verData
+       , Eq verData
+       , Serialise verData
+       , Typeable verData
+       )
+    => verData
+    -> (verData -> verData -> r)
+    -> VersionData ver verData r
+
+instance Show ver => Show (ProtocolVersion ver r) where
+  show (ProtocolVersion ver (VersionData verData _)) = 
+    "ProtocolVersion " ++ show ver ++ " " ++ show verData
+
+instance Eq (VersionData ver verData r) where
+  VersionData verData _ == VersionData verData' _ = verData == verData'
+
+-- |
+-- Get protocol version number out of @'ProtocolVersion'@.
+--
+protocolVersion :: ProtocolVersion ver r -> ver
+protocolVersion (ProtocolVersion ver _) = ver
+
+-- |
+-- Encode version data hidden in @'ProtocolVersion'@, encoding is done using
+-- the @'Serialise'@ constraint provided by @'VersionData'@ constructor.
+--
+encodeVersionData :: Enum ver => ProtocolVersion ver r -> Encoding
+encodeVersionData (ProtocolVersion _ (VersionData verData _)) =
+  encode verData
+
+-- |
+-- Decode remote version data using a local @'ProtocolVersion'@.
+--
+decodeVersionData :: (Enum ver, Show ver)
+                  => ProtocolVersion ver r
+                  -> Decoder s (ProtocolVersion ver r)
+decodeVersionData (ProtocolVersion ver (VersionData (_ :: verData) k)) =
+  ProtocolVersion ver . flip VersionData k <$> decode
+
+-- |
+-- Check if two @'ProtocolVersion'@ have the same version number @'ver'@.
+--
+protocolVersionEq :: Eq ver
+                  => ProtocolVersion ver r
+                  -> ProtocolVersion ver r
+                  -> Bool
+protocolVersionEq pv pv' = protocolVersion pv == protocolVersion pv'
+
+protocolVersionOrd :: Ord ver
+                   => ProtocolVersion ver r
+                   -> ProtocolVersion ver r
+                   -> Ordering
+protocolVersionOrd pv pv' = protocolVersion pv `compare` protocolVersion pv'
+
+-- |
+-- Sorted list of @'ProtocolVersion'@s.
+--
+data KnownVersions ver r where
+  KnownVersions :: [ProtocolVersion ver r] -> KnownVersions ver r
```

Comment:
Why not use a `Map` keyed on `ver`?

Reply:
It gets rid of the order brought about by a list, which ought to be completely irrelevant.